client/app/scripts/superdesk/editor/editor.js

```diff
@@ -550,16 +550,375 @@ function EditorService(spellcheck, $rootScope, $timeout) {
+                        }
+                    },
+                    retrieveEmbed:function() {
+                        // if it's an url, use embedService to retrieve the embed code
+                        var embedCode;
+                        if (_.startsWith(vm.input, 'http')) {
+                            embedCode = embedService.get(vm.input).then(function(data) {
+                                return {
+                                    body: data.html,
+                                    provider: data.provider_name
+                                };
+                            });
+                        // otherwise we use the content of the field directly
+                        } else {
+                            var embedType = 'Custom';
+                            // try to guess the provider of the custom embed
+                            if (vm.input.indexOf('twitter.com/widgets.js')) {
+                                embedType = 'Twitter';
+                            } else if (vm.input.indexOf('https://www.youtube.com')){
+                                embedType = 'YouTube';
+                            }
+                            embedCode = $q.when({
+                                body: vm.input,
+                                provider: embedType
+                            });
+                        }
+                        return embedCode;
+                    },
+                    updatePreview: function() {
+                        vm.retrieveEmbed().then(function(embed) {
+                            angular.element($element).find('.preview').html(embed.body);
+                        });
+                    },
+                    createFigureBlock: function(embedType, embedBody) {
+                        // create a new block containing the embed
+                        return vm.editorCtrl.insertNewBlockAfter(vm.blockBefore, {
+                            blockType: 'embed',
+                            embedType: embedType,
+                            body: [
+                                '<figure>',
+                                embedBody,
+                                '</figure>'
+                            ].join('\n')
+                        });
+                    },
+                    createBlockFromEmbed: function() {
+                        vm.retrieveEmbed().then(function(embed) {
+                            vm.createFigureBlock(embed.provider, embed.body);
+                            // close the addEmbed form
+                            vm.toggle(true);
+                        });
+                    },
+                    createBlockFromPicture: function(picture) {
+                        vm.createFigureBlock('Image', [
+                            '<img alt="' + picture.description + '" src="' + picture.renditions.viewImage.href + '">',
+                            '<figcaption>' + picture.headline + '</figcaption>'
```

Comment:
Just checking - is `picture.headline` properly escaped by this point?
(the same for its `description` and any other attributes)